### PR TITLE
docker_login: allow to function with credential helpers

### LINKED ISSUE #29579: Support for Logon to Amazon EC2 Container Registry
From @osiegmar on 2016-10-16T15:34:56Z
##### ISSUE TYPE

- Feature Idea
##### COMPONENT NAME

docker_login
##### ANSIBLE VERSION

```
2.1.1.0
```
##### SUMMARY

As far as I know, the only way to logon to (and pull docker images from / push to) [Amazon ECR](https://aws.amazon.com/de/ecr/) is via the shell module. It would be nice if the docker_login module supports logon to ECR so that further ansible docker tasks can directly work with ECR.

This is how I handle this currently:

```
- name: ECR login
  shell: "$(aws ecr get-login --region eu-central-1)"
- name: Pull image from ECR
  shell: "docker pull myid.dkr.ecr.eu-central-1.amazonaws.com/my-app:latest"
```

The output from `aws ecr get-login` is:

```
docker login -u AWS -p [VERY-LONG-PASSWORD] -e none https://myid.dkr.ecr.eu-central-1.amazonaws.com
```

The generated password is valid for 12 hours.

Of course, I could use awk to get the password from the output and then use docker_login. But the output format may change...

Copied from original issue: ansible/ansible-modules-core#5277

### LINKED ISSUE #32868: docker_login module returns failed=false even when login fails
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

`docker_login`

##### ANSIBLE VERSION

```
ansible 2.4.0.0
  config file = /Users/ssteiner/Documents/Repos/foo/ansible/ansible.cfg
  configured module search path = [u'/Users/foo/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python2.7/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.14 (default, Sep 25 2017, 09:53:22) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.37)]
```

##### CONFIGURATION

```
ANSIBLE_SSH_RETRIES(/Users/foo/Documents/Repos/foo/ansible/ansible.cfg) = 3
DEFAULT_HOST_LIST(/Users/foo/Documents/Repos/foo/ansible/ansible.cfg) = [u'/Users/foo/Documents/Repos/foo/ansible/inventory']
DEFAULT_ROLES_PATH(/Users/foo/Documents/Repos/foo/ansible/ansible.cfg) = [u'/Users/foo/Documents/Repos/foo/ansible/roles']
HOST_KEY_CHECKING(/Users/foo/Documents/Repos/foo/ansible/ansible.cfg) = False
```

##### OS / ENVIRONMENT

MacOS Sierra connecting to Ubuntu 16.04.03 and Docker Server Version: 17.09.0-ce

##### SUMMARY

When using `become: true`, the docker_login module returns `"failed" : false` even if bad credentials are passed and login to Docker Hub is unsuccessful.

##### STEPS TO REPRODUCE

Run the playbook below.

```yaml
---
- hosts: all
  become: true
  gather_facts: no
  vars_prompt:
    - name: hub_username
      prompt: "Enter your Docker Hub username"
      private: no
    - name: hub_email
      prompt: "Enter your Docker Hub email address"
      private: no
    - name: hub_password
      prompt: "Enter your Docker Hub password"
      private: yes
  tasks:
  - name: login to Docker Hub
    docker_login:
      username: "{{ hub_username }}"
      email: "{{ hub_email }}"
      password: "{{ hub_password }}"
    register: login_results
  - debug: 
      var: login_results
```

##### EXPECTED RESULTS

```
fatal: [node1]: FAILED! => {"changed": false, "failed": true, "msg": "Logging into https://index.docker.io/v1/ for user <redacted> failed - 401 Client Error: Unauthorized (\"Get https://registry-1.docker.io/v2/: unauthorized: incorrect username or password\")"}
```

##### ACTUAL RESULTS

```
ok: [node1] => {
    "login_results": {
        "changed": true,
        "failed": false,
        "login_result": {
            "email": "<redacted>",
            "password": "foo", 
            "serveraddress": "https://index.docker.io/v1/",
            "username": "<redacted>"
        }
    }
}
```
Coincidentally, it seems to echo the password back even though it should not.

```
ok: [node1] => {
    "login_results": {
        "changed": true,
        "failed": false,
        "login_result": {
            "email": "<redacted>",
            "password": "foo",
            "serveraddress": "https://index.docker.io/v1/",
            "username": "<redacted>"
        }
    }
}
```

### LINKED ISSUE #55738: Docker_login doesn't use Credentials store
##### SUMMARY
docker login encourages you using Credential store.
I use "D-Bus Secret Service".
docker_login module returns OK. But it writes credentials in `~/.docker/config.json` instead of using `docker-credential-secretservice`

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
docker_login

##### ANSIBLE VERSION

```paste below
ansible 2.7.5
  config file = /[REDACTED]/architecture/ansible.cfg
  configured module search path = ['/home/[USERNAME]/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /[REDACTED]/architecture/ansible-env/lib/python3.5/site-packages/ansible
  executable location = /[REDACTED]/architecture/ansible-env/bin/ansible
  python version = 3.5.3 (default, Sep 27 2018, 17:25:39) [GCC 6.3.0 20170516]
```

##### CONFIGURATION

```paste below
ANSIBLE_PIPELINING(/[REDACTED]/architecture/ansible.cfg) = True
DEFAULT_CALLBACK_WHITELIST(/[REDACTED]/architecture/ansible.cfg) = ['profile_tasks', 'timer']
DEFAULT_HOST_LIST(/[REDACTED]/architecture/ansible.cfg) = ['/[REDACTED]/architecture/ansible/inventory/local']
DEFAULT_MANAGED_STR(/[REDACTED]/architecture/ansible.cfg) = This file is managed by Ansible. Generated on  %Y-%m-%d %H:%M:%S by user {uid} on  %Y-%m-%d %H:%M:%S by user {uid} on host {host}
DEFAULT_ROLES_PATH(/[REDACTED]/architecture/ansible.cfg) = ['/[REDACTED]/architecture/ansible/roles']

```

##### OS / ENVIRONMENT
* PC DELL LATITUDE 7470
* Devuan GNU/Linux 2.0 (ascii) (debian 9 based distro)
* xfce4
* Docker version 18.09.5, build e8ff056dbc
* python virtualenv

##### STEPS TO REPRODUCE
* set up docker "secret servcie" store
add in ~/.docker/config.json
```json
"credsStore": "secretservice"
```
put in your $PATH the unarchived version of https://github.com/docker/docker-credential-helpers/releases/download/v0.6.0/docker-credential-secretservice-v0.6.0-amd64.tar.gz

```yaml
- block:
  - name: Login to a private docker registry
    docker_login:
      registry: "my-private.remote-docker-registry.host.example"
      username: username
      password: password
```

##### EXPECTED RESULTS

```bash
# RUN
docker-credential-secretservice list
# It should output the following
{"my-private.remote-docker-registry.host.example":"username"}

# RUN
cat ~/.docker/config.json
# It should output the following
{
        "auths": {
                "my-private.remote-docker-registry.host.example": {}
        },
        "HttpHeaders": {
                "User-Agent": "Docker-Client/18.09.5 (linux)"
        },
        "credsStore": "secretservice"
}

```

##### ACTUAL RESULTS

```paste below
# RUN
docker-credential-secretservice list
# It should output the following
{}

# RUN
cat ~/.docker/config.json
# It should output the following
{
     "HttpHeaders": {
          "User-Agent": "Docker-Client/18.09.5 (linux)"
     },
     "auths": {
          "my-private.remote-docker-registry.host.example": {
               "auth": "[REDACTED]",
               "email": null
          }
     },
     "credsStore": "secretservice"

}
```

### PR DESCRIPTION
**Pull request by felixfontein**
##### SUMMARY
Rebase of #61207 minus the unit tests from that PR (these require new dependencies, so they need to be rewritten or need approval of testing team for new dependency).

Fixes #55738
Fixes #32868
Potentially fixes #29579 (with https://github.com/awslabs/amazon-ecr-credential-helper)
Closes #61207

##### ISSUE TYPE
- Bugfix Pull Request
- New Module Pull Request

##### COMPONENT NAME
docker_login

### CODE REVIEW

**Comment by felixfontein** [comment]
CC @arlaneenalra

**Comment by felixfontein** [comment]
According to this, `email` has been deprecated by Docker: https://github.com/moby/moby/blob/master/api/types/auth.go#L9-L12

**Comment by felixfontein** [comment]
I think this fixes a bunch of long-standing open issues! I'll ask in the issues for people to test whether this helps them. Besides that, we have to find out whether this rewrite can count as a bugfix (will also depend on what users from the issues report) or a new feature. Bugfix will mean we can backport this to 2.9 or even 2.8, and new feature means this has to wait for 2.10 (probably half a year in the future, extrapolating from earlier release cycles).

**Comment by felixfontein** [comment]
Sorry, these commits went to the wrong branch...

File: lib/ansible/modules/cloud/docker/docker_login.py [hunk 1]
```diff
@@ -22,9 +22,10 @@
 version_added: "2.0"
 description:
     - Provides functionality similar to the "docker login" command.
-    - Authenticate with a docker registry and add the credentials to your local Docker config file. Adding the
-      credentials to the config files allows future connections to the registry using tools such as Ansible's Docker
-      modules, the Docker CLI and Docker SDK for Python without needing to provide credentials.
+    - Authenticate with a docker registry and add the credentials to your local Docker config file respectively the
+      credentials store associated to the registry. Adding the credentials to the config files resp. the credential
```
Comment (Andersson007) [inline lib/ansible/modules/cloud/docker/docker_login.py:26]: i don't use it, so only a small suggestion and question about the doc
```suggestion
      credentials store associated to the registry.
    - Adding the credentials to the config files resp. the credential
```
and ```resp. the``` is this an abbreviation?

Comment (felixfontein) [reply lib/ansible/modules/cloud/docker/docker_login.py:26]: I would keep it in one paragraph, as these two sentences belong together.

And `resp.` is an abbrevation for `respectively`.

Comment (Andersson007) [reply lib/ansible/modules/cloud/docker/docker_login.py:26]: Ok, thanks for clarifying! will know

File: lib/ansible/modules/cloud/docker/docker_login.py [hunk 2]
```diff
@@ -316,7 +449,7 @@ def main():
         registry_url=dict(type='str', default=DEFAULT_DOCKER_REGISTRY, aliases=['registry', 'url']),
         username=dict(type='str'),
         password=dict(type='str', no_log=True),
-        email=dict(type='str'),
+        email=dict(type='str', removedin_version='2.14'),
```
Comment (resmo) [inline lib/ansible/modules/cloud/docker/docker_login.py]: should be `removed_in_version`, see https://docs.ansible.com/ansible/latest/dev_guide/developing_program_flow_modules.html#removed-in-version

**Review by resmo — COMMENTED** [review]
I see some more params having this typo... currently in release. The question is how to proceed. change the version to 2.13 and backport to 2.9 would be my suggestion, but asking the release manager for this case @abadger 

(it would also have some docs, comments and code to update)

~~~diff
-        ), removedin_version='2.12'),
-        dockerfile=dict(type='str', removedin_version='2.12'),
+        ), removed_in_version='2.13'),
+        dockerfile=dict(type='str', removed_in_version='2.13'),
         force=dict(type='bool', removed_in_version='2.12'),
         force_source=dict(type='bool', default=False),
         force_absent=dict(type='bool', default=False),
         force_tag=dict(type='bool', default=False),
-        http_timeout=dict(type='int', removedin_version='2.12'),
+        http_timeout=dict(type='int', removed_in_version='2.13'),
         load_path=dict(type='path'),
         name=dict(type='str', required=True),
-        nocache=dict(type='bool', default=False, removedin_version='2.12'),
-        path=dict(type='path', aliases=['build_path'], removedin_version='2.12'),
-        pull=dict(type='bool', removedin_version='2.12'),
+        nocache=dict(type='bool', default=False, removed_in_version='2.13'),
+        path=dict(type='path', aliases=['build_path'], removed_in_version='2.13'),
+        pull=dict(type='bool', removed_in_version='2.13'),
         push=dict(type='bool', default=False),
         repository=dict(type='str'),
-        rm=dict(type='bool', default=True, removedin_version='2.12'),
+        rm=dict(type='bool', default=True, removed_in_version='2.13'),
         state=dict(type='str', default='present', choices=['absent', 'present', 'build']),
         tag=dict(type='str', default='latest'),
         use_tls=dict(type='str', choices=['no', 'encrypt', 'verify'], removed_in_version='2.11'),
-        buildargs=dict(type='dict', removedin_version='2.12'),
+        buildargs=dict(type='dict', removed_in_version='2.13'),
     )
 

~~~

File: lib/ansible/modules/cloud/docker/docker_login.py (hunk 2 above)
Comment (felixfontein) [reply lib/ansible/modules/cloud/docker/docker_login.py]: Thanks for spotting this, I've just fixed it here. For `docker_image`, I'll create a PR with some more text there :)

Comment (felixfontein) [reply lib/ansible/modules/cloud/docker/docker_login.py]: For anyone interested, the PR is #63345.

**Comment by felixfontein** [comment]
@Andersson007 @resmo any more comments?
@WojciechowskiPiotr can you take a look at this as well?

**Review by Andersson007 — APPROVED** [review]
@felixfontein , no, i haven't )

**Comment by felixfontein** [comment]
@deogracia @ntwrkguru @ssbarnea I would be happy to know if this PR fixes the problems you reported on `docker_login`.

**Comment by Grisu118** [comment]
I tested this branch regarding issue #64392. this also fixes the issue 👍

**Comment by felixfontein** [comment]
Rebased to resolve conflicts.

**Comment by felixfontein** [comment]
If nobody minds, I'll merge this PR this weekend.

**Comment by felixfontein** [comment]
@arlaneenalra thanks a lot for this rewrite! I'm sure it will improve the situation a lot!
@Andersson007 @resmo @Grisu118 thanks a lot for reviewing and testing this!

**Comment by arlaneenalra** [comment]
Glad to he of service! Sorry I ran out of time to do more with it.

**Pull request merged**